sw/host/opentitanlib/src/util/openocd.rs

```diff
@@ -0,0 +1,256 @@
+#[derive(Error, Debug, Deserialize, Serialize)]
+pub enum OpenOcdServerError {
+    #[error("OpenOCD server is not running")]
+    OpenOcdNotRunning,
+    #[error("Generic error {0}")]
+    Generic(String),
+}
+impl_serializable_error!(OpenOcdServerError);
+
+impl OpenOcdServer {
+    /// Create a new OpenOcdServer with the given JTAG options without starting OpenOCD.
+    pub fn new(opts: &JtagParams) -> Result<OpenOcdServer> {
+        Ok(OpenOcdServer {
+            opts: opts.clone(),
+            openocd_server_process: None.into(),
+            openocd_socket_stream: None.into(),
+        })
+    }
+
+    /// Spawn the OpenOCD server and connect to it.
+    fn start(&self, tap: JtagTap) -> Result<()> {
+        // Check if a server was already started.
+        if let Some(mut child) = self.openocd_server_process.take() {
+            // If it's still running, it won't have an exit status
+            if child.try_wait()?.is_none() {
+                self.openocd_server_process.replace(Some(child));
+                bail!("OpenOCD server already running");
+            }
+        }
+
+        let mut args = Vec::new();
+
+        // Pass the path to the adapter config file if given.
+        if let Some(cfg) = &self.opts.openocd_adapter_config {
+            args.extend(["-f", &cfg]);
+        }
+
+        // Because these commands are command-line args, they *must* end with
+        // semicolons to satisfy the TCL lexer. When a TCL program is read from
+        // a file, newlines are sufficient to separate statements.
+        args.extend(["-c", "adapter speed 200;"]);
+        args.extend(["-c", "transport select jtag;"]);
+        args.extend(["-c", "scan_chain;"]);
+
+        // Pass through the config for the chosen TAP
+        let target = match tap {
+            JtagTap::RiscvTap => &self.opts.openocd_riscv_target_config,
+            JtagTap::LcTap => &self.opts.openocd_lc_target_config,
+        };
+        if let Some(cfg) = &target {
+            args.extend(["-f", &cfg]);
+        }
+
+        // Use the default port explicity to avoid relying on it not changing
+        let port_cmd = format!("tcl_port {OPENOCD_TCL_PORT}");
+        args.extend(["-c", port_cmd.as_str()]);
+        args.extend(["-c", "init;"]);
+
+        log::info!("CWD: {:?}", std::env::current_dir());
+        log::info!(
+            "Spawning OpenOCD: {:?} {:?}",
+            self.opts.openocd,
+            args.join(" ")
+        );
+
+        // TODO: forward output/stderr to loggin using a worker, like for verilator
+        let child = Command::new(&self.opts.openocd)
+            .args(args)
+            .stdin(Stdio::null())
+            .stdout(Stdio::inherit())
+            .stderr(Stdio::inherit())
+            .spawn()
+            .context("failed to spawn openocd")?;
+        self.openocd_server_process.replace(Some(child));
+
+        // Give OpenOCD time to start up
+        thread::sleep(Duration::from_millis(1000));
```

Comment:
I am not a huge fan of hardcoded delays like that, they can lead to really difficult errors to debug. Maybe it would be better to retry several times with a smallish timeout (100ms) and bail out after 5s or a user configurable timeout? That's assuming we can get some reasonable errors out of `TcpStream:connect`

Reply:
Since the PR is getting long, we have added this to #17490 and will do it properly in another PR.